feat: add `math/base/special/tribonaccif`

Neerajpathak07 opened the pull request:
Progresses #649.

## Description

> What is the purpose of this pull request?

This pull request:

-   adds `math/base/special/tribonaccif` which is the single-precision version of `tribonacci`

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   progresses #649 

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-  [x] Read, understood, and followed the [contributing guidelines](https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md)

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

Neerajpathak07 committed:
feat: add-tribonaccif

Neerajpathak07 committed:
chore: update C implementation

Neerajpathak07 committed:
chore: update manifest

Neerajpathak07 commented on lib/node_modules/@stdlib/math/base/special/tribonaccif/src/main.c:
```diff
@@ -0,0 +1,75 @@
+#include "stdlib/constants/float32/max_safe_nth_tribonacci.h"
+#include "stdlib/math/base/assert/is_nanf.h"
+
+static const int32_t tribonaccif_value[ 31 ] = {
```
went with 31 values here.

Neerajpathak07 commented on lib/node_modules/@stdlib/math/base/special/tribonaccif/src/main.c:
```diff
@@ -0,0 +1,75 @@
+* // returns 0
+*/
+float stdlib_base_tribonaccif( const int32_t n ) {
+	if ( n < 0 || n > STDLIB_CONSTANT_FLOAT32_MAX_SAFE_NTH_TRIBONACCI || n == 3.14 ) { // eslint-disable-line max-len
+		return 0.0 / 0.0; // NaN
+	}
+	if ( stdlib_base_is_nanf( n ) ){
+		return 0.0 / 0.0; // NaN
+	}
```
Also tried to explicitly handle the failing cases here. LMK if I am thinking straight.
cc:- @kgryte @Planeshifter

Neerajpathak07 committed:
refactor: C implementation

Neerajpathak07 commented:
cc:- @kgryte @Planeshifter LMK what am I missing here!!

Neerajpathak07 committed:
refactor: C implementation

Neerajpathak07 committed:
chore: update tests

Neerajpathak07 commented:
@Planeshifter Hey could I pls know why the `potential Dupliate` label was added as this package had not been worked upon and I did specifically add `max_nth_safe_tribonacci` pkg for `float32` to implement `tribonaccif`, also for other `constants/float32` packages?

kgryte commented:
@Neerajpathak07 Simple, because you consistently use `resolves` incorrectly. No one of your PR "resolves" or "closes" #649. Instead, you should be using "progresses". Only if your PR actually completely addresses an issue should you be using one of those keywords.

Because you consistently use `resolves #649`, your PRs are being flagged as potential duplicates.

See also https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/using-keywords-in-issues-and-pull-requests.

Neerajpathak07 commented:
@kgryte ahh I see alright then I shall update all the other PR's for this.
Thank you soo much for addressing this.

Neerajpathak07 committed:
chore: update README

Neerajpathak07 commented on lib/node_modules/@stdlib/math/base/special/tribonaccif/src/main.c:
```diff
@@ -0,0 +1,71 @@
+	2555757,
+	4700770,
+	8646064,
+	15902591
```
Have taken values till here as the `FLOAT32_MAX_SAFE_INTEGER` is 16777215 so any value beyond this just returns `NaN`

Neerajpathak07 commented on lib/node_modules/@stdlib/math/base/special/tribonaccif/src/main.c:
```diff
@@ -0,0 +1,71 @@
+* // returns 0f
+*/
+float stdlib_base_tribonaccif( const int32_t n ) {
+	if ( n < 0 || n > STDLIB_CONSTANT_FLOAT32_MAX_SAFE_NTH_TRIBONACCI ) {
+		return 0.0 / 0.0; // NaN
+	}
```
Haven't added cases for `is_nanf` and to check if `n` is int. As while operating on `int32_t` variables we don't want to return `NaN` for integers and `NaN` itself.

Neerajpathak07 commented:
@kgryte Kindly review this and LMK if I am missing something!!

Neerajpathak07 commented on lib/node_modules/@stdlib/math/base/special/tribonaccif/lib/native.js:
```diff
@@ -0,0 +1,74 @@
+* var y = tribonaccif( 5 );
+* // returns 4
+*
+* @example
+* var y = tribonaccif( 6 );
+* // returns 7
+*
+* @example
+* var y = tribonaccif( -1 );
+* // returns NaN
+*/
```
Also removed the example for testing `tribonaccif` value for `3.14` integer. As it just truncates the value and return s 0 which is undesirable.

Neerajpathak07 committed:
chore: update repl input value

Neerajpathak07 committed:
chore: refactor READEM file

Neerajpathak07 commented:
@kgryte seems like Gunj is AFK for a while for the time-being could you look into it?

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/tribonaccif/README.md:
```diff
@@ -0,0 +1,254 @@
+
+# Tribonaccif
+
+> Compute the nth [Tribonacci number][tribonacci-number] for [single-precision floating-point number][ieee754].
```
```suggestion
> Compute the nth [Tribonacci number][tribonacci-number] as a [single-precision floating-point number][ieee754].
```

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/tribonaccif/README.md:
```diff
@@ -0,0 +1,254 @@
+
+#### tribonaccif( n )
+
+Computes the nth [Tribonacci number][tribonacci-number] for [single-precision floating-point number][ieee754].
```
```suggestion
Computes the nth [Tribonacci number][tribonacci-number] as a [single-precision floating-point number][ieee754].
```

gunjjoshi replied:
@Neerajpathak07 Can you update this here and elsewhere? Because, we are not pointing `n` to be in single-precision, but we are referring to the nth tribonacci number to be in single-precision.

Neerajpathak07 replied:
sure

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/tribonaccif/README.md:
````diff
@@ -0,0 +1,254 @@
+If `n > 30`, the function returns `NaN`, as larger [Tribonacci numbers][tribonacci-number] cannot be safely represented in [single-precision floating-point format][ieee754].
+
+```javascript
+var v = tribonaccif( 32 );
````
```suggestion
var v = tribonaccif( 31 );
```
Even though 32 is also fine, let's keep the closest value to the limit of 30, i.e., 31 here, similar to what we have in the double-precision equivalent, `tribonacci`.

Neerajpathak07 replied:
Ok np I'll update it in the `repl` also.

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/tribonaccif/README.md:
```diff
@@ -0,0 +1,254 @@
+
+#### stdlib_base_tribonaccif( n )
+
+Computes the nth [Tribonacci number][tribonacci-number] [single-precision floating-point number][ieee754].
```
```suggestion
Computes the nth [Tribonacci number][tribonacci-number] [single-precision floating-point number][ieee754].
```
This must be updated.

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/tribonaccif/README.md:
````diff
@@ -0,0 +1,254 @@
+Computes the nth [Tribonacci number][tribonacci-number] [single-precision floating-point number][ieee754].
+
+```c
+float out = stdlib_base_tribonaccif( 0f );
+// returns 0f
+
+out = stdlib_base_tribonaccif( 1f );
+// returns 0f
````
```suggestion
float out = stdlib_base_tribonaccif( 0 );
// returns 0

out = stdlib_base_tribonaccif( 1 );
// returns 0
```

gunjjoshi replied:
@Neerajpathak07 We don't need to add the `f` suffixes after integers here.

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/tribonaccif/README.md:
```diff
@@ -0,0 +1,254 @@
+[tribonacci-number]: https://en.wikipedia.org/wiki/Generalizations_of_Fibonacci_numbers#Tribonacci_numbers
+
+[ieee754]: https://en.wikipedia.org/wiki/IEEE_754-1985
+
```
```suggestion

<!-- <related-links> -->

<!-- </related-links> -->

```

gunjjoshi replied:
Missing section for related links.

gunjjoshi commented:
@Neerajpathak07 Looks like this PR is missing `benchmark/c/native` directory and the files inside it.

Neerajpathak07 commented:
@gunjjoshi Let me get back to you with the updated changes!!

Neerajpathak07 committed:
chore: changes from code review

Neerajpathak07 committed:
chore: adding sections for related packages

anandkaranubc commented:
/stdlib merge

stdlib-bot committed:
Merge remote-tracking branch 'upstream/develop' into add-tribonaccif

anandkaranubc committed:
chore: code review

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: passed
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte commented on lib/node_modules/@stdlib/math/base/special/tribonaccif/README.md:
```diff
@@ -0,0 +1,246 @@
+
+-->
+
+# Tribonaccif
```
```suggestion
# tribonaccif
```

Either needs to be the alias or a title. E.g., here either "Tribonacci" (title) or an alias "tribonaccif". No mixing the two.

anandkaranubc replied:
Noted! Thank you!

kgryte committed:
docs: update heading

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/math/base/special/tribonaccif/benchmark/c/benchmark.c:
```diff
@@ -0,0 +1,160 @@
+* @param n   Tribonacci number to compute
+* @return    Tribonacci number
+*/
+int tribonaccif( int n ) {
```
```suggestion
static int tribonaccif( int n ) {
```

kgryte committed:
bench: add `static` qualifier

Signed-off-by: Athan <kgryte@gmail.com>

kgryte reviewed (approved):
LGTM
